feat(discover-snql): Add the ability to use projects

wmak opened the pull request:
- This adds our project slug transforms to the snql query builder
- This also adds a `wip_snql_query` function to discover so that we can do integration tests with the snql queries as well to make sure our created queries can run sucessfully
  - Went with a function containing `wip` so its clearer that this isn't a function that's ready to use yet
- This also fixed some issues with the QueryFields.orderby
- Switched tests to [assertCountEqual](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual), since we don't care about the order of conditions

wmak committed:
feat(discover-snql): Add the ability to use projects

- This adds our existing project slug transforms to the snql query
  builder
- This also adds a `wip_snql_query` function to discover so that we can
  do integration tests with the snql queries as well to make sure our
  created queries can run sucessfully
  - Went with a function containing `wip` so its clearer that this isn't
    a function that's ready to use yet
- This also fixed some issues with the QueryFields.orderby

wmak commented on src/sentry/search/events/base.py:
```diff
@@ -15,7 +15,7 @@ def __init__(
         dataset: Dataset,
         params: ParamsType,
-        orderby: Optional[List[str]] = None,
+        orderby: Optional[Union[List[str], str]] = None,
```
Support `orderby="project"` but also `orderby=["project", "message"]`

wmak committed:
ref: extra pass

wmak commented on src/sentry/search/events/fields.py:
```diff
@@ -1840,13 +1840,51 @@ class QueryFields(QueryBase):
+
+        projects = Project.objects.filter(id__in=project_ids).values("slug", "id")
+
+        return SnqlFunction(
+            "transform",
+            [
+                self.column("project.id"),
+                [project["id"] for project in projects],
+                [project["slug"] for project in projects],
+                "",
+            ],
+            project_key,
+        )
```
This is so much cleaner than the previous transform, no `array` functions, no random toString on the project.id 😭

wmak commented on src/sentry/search/events/fields.py:
```diff
@@ -1840,13 +1840,51 @@ class QueryFields(QueryBase):
+
+    def project_slug_transform(self, project_key: str) -> SnqlFunction:
+        """When project is a selected column we need to create a transform to turn them back into slugs"""
+        project_ids = {
+            project_id
+            for project_id in self.params.get("project_id", [])
+            if isinstance(project_id, int)
+        }
+
+        # Try to reduce the size of the transform by using any existing conditions on projects
+        if len(self.projects_to_filter) > 0:
+            project_ids &= self.projects_to_filter
```
The existing transform logic scans through conditions to figure out what to exclude, I've changed the class to construct the `projects_to_filter` during condition construction instead

wmak commented on src/sentry/search/events/filter.py:
```diff
@@ -851,6 +852,50 @@ def _environment_filter_converter(
+
+        if search_filter.operator == "=" and value == "":
+            raise InvalidSearchQuery(
+                'Cannot query for has:project or project:"" as every event will have a project'
```
reworded from the original message of `Invalid query for 'has' search: 'project' cannot be empty.`

wmak commented on src/sentry/search/events/filter.py:
```diff
@@ -851,6 +852,50 @@ def _environment_filter_converter(
+            )
+            if converted_filter:
+                if search_filter.operator in EQUALITY_OPERATORS:
+                    self.projects_to_filter.update(project_ids)
```
new way of doing the `project_to_filter` method, afaik we don't do INEQUALITY here, since more complicated boolean searches can result in a project being added back in.

Zylphrex replied:
This will convert something like `project:a project:b` into an `OR` rather than an `AND` condition right?

wmak replied:
Nah so the actual conditions will still be correct (see L887 the converted_filter will add `WHERE project = 1` correctly for example) this `projects_to_filter` list is cause we need to send a root(base?) condition on project_id
eg.
params[project_id] = [1,2,3]
condition = (project:1 and transaction:foo) or (project:2 and transaction:bar)

results in
select...where project_ids IN [1,2] and ((project=1 and txn=foo) or (project=2 and txn=bar))


Though there's currently a bug with this that's in the backlog(ie. i don't want to fix as part of the migration) something like this:
eg.
params[project_id] = [1,2,3]
condition = (project:1 and transaction:foo) or (transaction:bar)

results in
select...where project_ids IN [1] and ((project=1 and txn=foo) or (txn=bar))

shruthilayaj commented on src/sentry/search/events/filter.py:
```diff
@@ -851,6 +852,50 @@ def _environment_filter_converter(
+        """
+        value = search_filter.value.value
+
+        if search_filter.operator == "=" and value == "":
```
Should we have an enum for search filter oeprators?

shruthilayaj reviewed (approved):
Just some nits/questions for my benefit. LGTM.

Zylphrex commented on src/sentry/search/events/fields.py:
```diff
@@ -1869,13 +1907,21 @@ def resolve_field(self, field: str) -> Union[SnqlColumn, SnqlFunction]:
+            resolved_orderby = self.column(bare_orderby)
+            direction = Direction.DESC if orderby.startswith("-") else Direction.ASC
+
+            for selected_column in self.columns:
+                if isinstance(selected_column, SnqlColumn) and selected_column == resolved_orderby:
+                    validated.append(OrderBy(selected_column, direction))
+                    continue
+                elif (
+                    isinstance(selected_column, SnqlFunction)
+                    and selected_column.alias == bare_orderby
+                ):
+                    validated.append(OrderBy(selected_column, direction))
+                    continue
```
Nit: this nested loop is now `O(n^2)`, can this be flattened with 2 sets, one with the `SnqlColumn`s and `SnqlFunction`?

Zylphrex reviewed (approved):
Just a few minor things that we can revisit and iterate on, but nice! This is looking awesome

wmak committed:
ref: Addressing PR comments

- Caching project slug query so we only do it once
- Updating docstring format
- Using Op(search_filter.operator) instead of checking the literal value

wmak commented on src/sentry/search/events/filter.py:
```diff
@@ -851,6 +852,50 @@ def _environment_filter_converter(
+            for p in Project.objects.filter(
+                id__in=self.params.get("project_id", []), slug__in=slugs
+            )
+        }
```
As discussed in slack
- lol it does
- Going to cache this query on the QueryBuilder object so we don't make >1 call

wmak committed:
Merge branch 'master' into wmak/feat/discover-snql-projects
